lib/chef/provider/windows_task.rb

```diff
@@ -201,308 +224,303 @@ def action_disable
       end
 
-      def convert_system_date_format_to_ruby_long_date
-        date_format = get_system_short_date_format.dup
-        date_format.sub!("MMM", "%m")
-        common_date_format_conversion(date_format)
-        date_format.sub!("yy", "%Y")
-        date_format
+      def convert_in_minutes(frequency_modifier)
+        frequency_modifier * 60
       end
 
-      def convert_system_date_format_to_ruby_date_format
-        date_format = get_system_short_date_format.dup
-        date_format.sub!("MMM", "%b")
-        common_date_format_conversion(date_format)
-        date_format.sub!("yy", "%y")
-        date_format
-      end
+      #TODO : Try to optimize this method
+      # known issue : Since start_day and time is not mandatory while updating weekly frequency for which start_day is not mentioned by user idempotency
+      # is not gettting maintained as new_resource.start_day is nil and we fetch the day of week from start_day to set and its currently coming as nil and don't match with current_task
+      def task_needs_update?(task)
+        flag = false
+        current_task_trigger = task.trigger(0)
+        new_task_trigger = trigger
 
-      def common_date_format_conversion(date_format)
-        date_format.sub!("dd", "d")
-        date_format.sub!("d", "%d")
-        date_format.sub!("MM", "%m")
-        date_format.sub!("M", "%m")
-        date_format.sub!("yyyy", "%Y")
-      end
+        if new_resource.frequency == :on_idle
+          # Converting here new_resource.idle_time in seconds to compare
+          flag =  true if ISO8601::Duration.new(task.idle_settings[:idle_duration]) != ISO8601::Duration.new(new_resource.idle_time*60)
+        end
 
-      def get_system_short_date_format
-        return @system_short_date_format if @system_short_date_format
-        Chef::Log.debug "Finding system date format"
-        task_script = <<-EOH
-          [Console]::OutputEncoding = [Text.UTF8Encoding]::UTF8
-          [Globalization.Cultureinfo]::CurrentCulture.DateTimeFormat.ShortDatePattern
-        EOH
-        @system_short_date_format = powershell_out(task_script).stdout.force_encoding("UTF-8").gsub(/[\s+\uFEFF]/, "")
-        @system_short_date_format
+        flag = true if start_day_updated?(current_task_trigger, new_task_trigger) == true ||
+        start_time_updated?(current_task_trigger, new_task_trigger) == true ||
+        current_task_trigger[:trigger_type] != new_task_trigger[:trigger_type] ||
+        current_task_trigger[:type] != new_task_trigger[:type] ||
+        current_task_trigger[:random_minutes_interval].to_i != new_task_trigger[:random_minutes_interval].to_i ||
+        current_task_trigger[:minutes_interval] != new_task_trigger[:minutes_interval] ||
+        task.account_information != new_resource.user ||
+        task.max_run_time != new_resource.execution_time_limit.to_i * 1000 ||
+        task.application_name != new_resource.command ||
+        task.principals[:run_level] != run_level
```

Comment:
L#298 - L#312,
all the conditions may be checked by using &&
In case of any false, it will be returned then and there
Hence 'flag' variable may not be required